File: firmware/velocity_jig/web_tool/excitation.js
// Excitation generators, one per protocol the runbook calls for.
//
// A program is `{durationS, at(t) -> {linear, angular, label}}`. Piecewise
// protocols also expose `segments` so the UI can draw them. The player samples
// at 50 Hz; nothing here knows about serial ports.
//
// Two conventions worth stating once:
//
// - Commands are normalized to [-1, 1]. trainer.js maps to the OpenTX range.
// - `shuttle` alternates the sign of successive repetitions so the robot ends
//   up near where it started. It is the cheapest way to fit a step battery into
//   a short board, because it costs nothing in the fit: a reverse step is a
//   step, and E8 wants both directions anyway.

import { PLANT, MIN_HOLD_S, MIN_COAST_S } from './plant.js';

/** Every generator clamps its hold here, so no caller can ask for a stub run. */
const hold = (s) => Math.max(MIN_HOLD_S, s ?? MIN_HOLD_S);

function piecewise(segments, label = '') {
    const durationS = segments.length ? segments[segments.length - 1].t1 : 0;
    return {
        label,
        durationS,
        segments,
        at(t) {
            for (const s of segments) {
                if (t >= s.t0 && t < s.t1) return s;
            }
            return { linear: 0, angular: 0, label: 'done' };
        },
    };
}

function push(segs, dur, linear, angular, label) {
    const t0 = segs.length ? segs[segs.length - 1].t1 : 0;
    segs.push({ t0, t1: t0 + dur, linear, angular, label });
    return segs;
}

/** Settle time to let a coast finish before the next repetition. */
function coastTime(p = PLANT) {
    return Math.max(MIN_COAST_S, p.delayS + 4 * p.tauDecel + 0.3);
}

/**
 * Scale an amplitude set so its largest magnitude lands on the cap.
 *
 * Scaling, not clamping. Clamping flattens every over-cap level onto the same
 * value, which turns four distinct steps into two and silently destroys the
 * coverage the experiment was recording. Scaling keeps the levels distinct and
 * their ratios intact, which is what a regression through them needs.
 *
 * Under-cap sets are returned untouched: the cap is a ceiling, not a target.
 */
export function capAmplitudes(amplitudes, cap) {
    if (!cap || cap <= 0) return amplitudes;
    const peak = Math.max(...amplitudes.map(Math.abs));
    if (peak <= cap) return amplitudes;
    const k = cap / peak;
    return amplitudes.map((a) => Math.round(a * k * 1000) / 1000);
}

/**
 * E7, E10. Staircase from `from` to `to` in `step`, holding each level.
 *
 * No coast between levels: the staircase is continuous by design, because the
 * deadzone is the level at which motion first appears and returning to zero
 * between levels would double the run length for nothing.
 */
export function staircase({
    channel = 'linear',
    from = 0.01,
    to = 0.1,
    step = 0.01,
    holdS = MIN_HOLD_S,
    sign = 1,
}) {
    const segs = [];
    const h = hold(holdS);
    for (let a = from; a <= to + 1e-9; a += step) {
        const v = Math.round(a * 1000) / 1000;
        push(
            segs,
            h,
            channel === 'linear' ? sign * v : 0,
            channel === 'angular' ? sign * v : 0,
            `${channel} ${(sign * v).toFixed(2)}`,
        );
    }
    push(segs, coastTime(), 0, 0, 'coast');
    return piecewise(segs, `${channel} staircase ${from}-${to}`);
}

/**
 * E8, E11. Amplitude steps with a full coast between each.
 *
 * `holdS` is the knob that trades floor for fit quality, and MIN_HOLD_S is where
 * it stops. Two seconds is over six time constants, which reaches the plateau
 * with room to spare, so max speed reads straight off the trace instead of
 * coming out of a fitted model. Shorter holds are what stage 2 ran, and stage 2
 * could not fit the forward accel constant.
 */
export function steps({
    channel = 'linear',
    amplitudes = [0.25, 0.5, 0.75, 1.0],
    holdS = MIN_HOLD_S,
    sign = 1,
    shuttle = false,
    scale = 1.0,
}) {
    const segs = [];
    const h = hold(holdS);
    let dir = sign;
    for (const a of amplitudes) {
        const v = a * scale * dir;
        push(
            segs,
            h,
            channel === 'linear' ? v : 0,
            channel === 'angular' ? v : 0,
            `${channel} ${v.toFixed(2)}`,
        );
        push(segs, coastTime(), 0, 0, 'coast');
        if (shuttle) dir = -dir;
    }
    return piecewise(segs, `${channel} steps${shuttle ? ' (shuttle)' : ''}`);
}

/** E9. Dedicated coast tails: hold, then drop to zero instantly. */
export function coastTails({
    amplitude = 0.6,
    holdS = MIN_HOLD_S,
    reps = 10,
    sign = 1,
    shuttle = false,
}) {
    const segs = [];
    const h = hold(holdS);
    let dir = sign;
    for (let i = 0; i < reps; i++) {
        push(segs, h, amplitude * dir, 0, `hold ${(amplitude * dir).toFixed(2)}`);
        push(segs, coastTime() + 0.5, 0, 0, 'coast to standstill');
        if (shuttle) dir = -dir;
    }
    return piecewise(segs, `coast tails x${reps}`);
}

/** E5. Pure spins at increasing yaw rate, encoder attached and scrubbing. */
export function spins({ amplitudes = [0.15, 0.25, 0.35, 0.45], holdS = 4.0, sign = 1, angularCap = null }) {
    const segs = [];
    // E5 finds the rate limit, so it has to approach it, but the runbook caps
    // the top rate and the wheel is scrubbing the whole time.
    for (const a of capAmplitudes(amplitudes, angularCap)) {
        push(segs, hold(holdS), 0, a * sign, `spin ${(a * sign).toFixed(2)}`);
        push(segs, MIN_COAST_S, 0, 0, 'settle');
    }
    return piecewise(segs, 'lever-arm spins');
}

/**
 * E13. Coupling grid, linear crossed with angular.
 *
 * Cells are ordered so the linear command alternates sign when shuttling: a
 * grid at full linear amplitude is sixteen steps of floor otherwise.
 */
export function couplingGrid({
    linear = [0.25, 0.5, 0.75, 1.0],
    angular = [0, 0.1, 0.2, 0.3],
    holdS = MIN_HOLD_S,
    shuttle = false,
    scale = 1.0,
    angularCap = 1.0,
}) {
    const segs = [];
    const h = hold(holdS);
    let dir = 1;
    const cells = [];
    const angScaled = capAmplitudes(angular, angularCap);
    for (const l of linear) for (const a of angScaled) cells.push([l, a]);
    for (const [l, a] of cells) {
        const lin = l * scale * dir;
        const ang = a;
        push(segs, h, lin, ang, `lin ${lin.toFixed(2)} ang ${ang.toFixed(2)}`);
        push(segs, coastTime(), 0, 0, 'coast');
        if (shuttle) dir = -dir;
    }
    return piecewise(segs, `coupling grid ${linear.length}x${angular.length}`);
}

/**
 * E14, E15, E16. Maximal-length LFSR, deterministic from the seed.
 *
 * A 60 ms bit period covers roughly 0.2 to 8 Hz, which brackets the 58-78 ms
 * time constants and the 33 ms camera frame period. Seeds are logged so a
 * repetition can be replayed exactly or deliberately varied.
 */
export function prbs({
    channels = ['linear'],
    amplitude = 0.6,
    bitMs = 60,
    durationS = 30,
    seed = 0xace1,
    angularAmplitude = null,
    zeroMean = true,
}) {
    const bits = Math.ceil((durationS * 1000) / bitMs);
    const seq = {};
    let s = seed >>> 0 || 0xace1;
    for (const ch of channels) {
        const out = new Int8Array(bits);
        for (let i = 0; i < bits; i++) {
            // 16-bit maximal LFSR, taps 16,14,13,11
            const bit = ((s >> 0) ^ (s >> 2) ^ (s >> 3) ^ (s >> 5)) & 1;
            s = ((s >> 1) | (bit << 15)) & 0xffff;
            out[i] = bit ? 1 : -1;
        }
        seq[ch] = out;
        s = (s * 1103515245 + 12345) & 0xffff || 0xbeef; // decorrelate channels
    }
    const angAmp = angularAmplitude ?? amplitude;
    return {
        label: `PRBS ${channels.join('+')} ${bitMs} ms`,
        durationS,
        segments: null,
        seed,
        at(t) {
            const i = Math.min(bits - 1, Math.floor((t * 1000) / bitMs));
            const lin = seq.linear ? seq.linear[i] * amplitude : 0;
            const ang = seq.angular ? seq.angular[i] * angAmp : 0;
            return {
                linear: zeroMean ? lin : (lin + amplitude) / 2,
                angular: ang,
                label: `prbs bit ${i}`,
            };
        },
    };
}

/**
 * E17. Logarithmic chirp, 0.2 to 8 Hz.
 *
 * Phase is integrated in closed form so the sweep is continuous. A first-order
 * model with tau = 58 ms has its corner near 2.7 Hz, and the chirp should show
 * that corner where the step fit says it is.
 */
export function chirp({ channel = 'linear', f0 = 0.2, f1 = 8, amplitude = 0.4, durationS = 30 }) {
    const k = Math.log(f1 / f0) / durationS;
    return {
        label: `chirp ${channel} ${f0}-${f1} Hz`,
        durationS,
        segments: null,
        at(t) {
            const phase = (2 * Math.PI * f0 * (Math.exp(k * t) - 1)) / k;
            const v = amplitude * Math.sin(phase);
            return {
                linear: channel === 'linear' ? v : 0,
                angular: channel === 'angular' ? v : 0,
                label: `${(f0 * Math.exp(k * t)).toFixed(2)} Hz`,
            };
        },
    };
}

/** E18, E19. The human drives; the tool only logs what the sticks are doing. */
export function manual({ durationS = 60 }) {
    return {
        label: 'operator driving',
        durationS,
        segments: null,
        manual: true,
        at() {
            return { linear: 0, angular: 0, label: 'operator' };
        },
    };
}

/**
 * Predicted floor a program needs, by integrating the plant against it.
 *
 * Closed-form distance works for a single step. It does not work for a shuttle,
 * a PRBS, or a chirp, and those are exactly the protocols whose footprint is
 * hard to guess. So the whole program gets run through a first-order sim with
 * transport delay and separate accel and decel constants, and the answer is the
 * span between the furthest points reached in each direction.
 */
export function predictFootprint(program, p = PLANT, dt = 0.005) {
    const n = Math.ceil((program.durationS + 2.0) / dt);
    const delaySteps = Math.round(p.delayS / dt);
    const cmdHistory = [];
    let v = 0;
    let x = 0;
    let heading = 0;
    let minX = 0;
    let maxX = 0;
    let peak = 0;
    let peakYaw = 0;

    for (let i = 0; i < n; i++) {
        const t = i * dt;
        const c = t < program.durationS ? program.at(t) : { linear: 0, angular: 0 };
        cmdHistory.push(c);
        const applied = cmdHistory[Math.max(0, cmdHistory.length - 1 - delaySteps)];

        const target = applied.linear * (applied.linear >= 0 ? p.vSsFwd : p.vSsRev);
        const tau = Math.abs(target) > Math.abs(v) ? p.tauAccel : p.tauDecel;
        v += ((target - v) / tau) * dt;

        const w = applied.angular * p.wSs;
        heading += w * dt;
        x += v * dt;
        minX = Math.min(minX, x);
        maxX = Math.max(maxX, x);
        peak = Math.max(peak, Math.abs(v));
        peakYaw = Math.max(peakYaw, Math.abs(w));
    }
    return {
        spanM: maxX - minX,
        forwardM: maxX,
        backwardM: -minX,
        endM: x,
        peakSpeed: peak,
        peakYawRate: peakYaw,
        headingTurns: heading / (2 * Math.PI),
    };
}

/**
 * Peak combined demand a program asks for, and where it asks for it.
 *
 * The transmitter spends a single budget across both axes (`|linear| + |angular|
 * <= 1`, angular first), so a cell that asks for more comes back with its linear
 * term cut. That is what the robot does in a match and the command log records
 * the reduced value, but a grid whose top row quietly loses a third of its
 * linear command is a grid that measured something else. Worth saying before the
 * run, not after.
 */
export function peakDemand(program, dt = 0.01) {
    let peak = 0;
    let label = '';
    for (let t = 0; t < program.durationS; t += dt) {
        const c = program.at(t);
        const d = Math.abs(c.linear ?? 0) + Math.abs(c.angular ?? 0);
        if (d > peak) {
            peak = d;
            label = c.label ?? '';
        }
    }
    return { peak, label, saturates: peak > 1 + 1e-9 };
}

/**
 * Fit a program to a budget by scaling amplitude, retrying until it fits.
 *
 * Amplitude is the last knob for a reason (see plant.js), so `build` should
 * already have shortened the dwell and enabled shuttling before this runs.
 */
export function fitToBudget(build, budgetM, p = PLANT) {
    let scale = 1.0;
    for (let i = 0; i < 12; i++) {
        const program = build(scale);
        const fp = predictFootprint(program, p);
        if (fp.spanM <= budgetM || scale < 0.06) {
            return { program, footprint: fp, scale, fits: fp.spanM <= budgetM };
        }
        scale *= Math.max(0.5, (budgetM / fp.spanM) * 0.95);
    }
    const program = build(scale);
    return { program, footprint: predictFootprint(program, p), scale, fits: false };
}

File: playground/calibration/calib_lib/excitation.py
"""Excitation waveforms for velocity jig runs, declared in TOML.

This replaces the web tool's `excitation.js` plus its hard-coded 21-entry experiment
catalog. The catalog was the runbook compiled into JavaScript, so changing an amplitude or
trying a new excitation meant editing code. Here a waveform is a TOML table:

    [waveform.lin_step_full]
    kind = "step"
    channel = "linear"
    role = "fit"
    amplitudes = [0.25, 0.5, 0.75, 1.0]
    hold_s = 2.0

`kind`, `channel` and `role` are also what the fit routes on, so a new waveform reaches the
right fit without touching the fitter.

Commands are normalized to [-1, 1] and dimensionless. Mapping to the radio's integer range
is the transmitter link's job, and deadzone compensation is deliberately not applied
anywhere: the physical deadzone is one of the things being measured.
"""

from __future__ import annotations

import math
import sys
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Callable, Sequence

from auto_battlebot.velocity_jig import CHANNELS, ROLES, WAVEFORM_KINDS, ProtocolSegment

if sys.version_info >= (3, 11):
    import tomllib
else:
    import tomli as tomllib

# A held command shorter than this cannot show a plateau to read a steady state off. Stage 2
# held roughly a fifth of a second, which sat under the rise, and no maximum speed could be
# read from it. The floor is applied to every generated hold rather than trusted to config.
MIN_HOLD_S = 0.5
# Time to let the robot come to rest between cells, so each step's decay is measured from a
# standstill rather than from whatever the previous cell left behind.
MIN_COAST_S = 1.5

DEFAULT_RATE_HZ = 50.0


@dataclass
class Program:
    """A playable excitation.

    `segments` is present for piecewise programs and None for continuous ones (chirp, sine,
    triangle, PRBS, manual), where there is no cell structure to slice by. `at(t)` is the
    authority either way: the player samples it, and segments exist so the fit can label
    samples by phase.
    """

    name: str
    kind: str
    channel: str
    role: str
    duration_s: float
    at: Callable[[float], tuple[float, float]]
    segments: list[ProtocolSegment] | None = None
    label: str = ""
    params: dict[str, Any] = field(default_factory=dict)

    def sample(self, rate_hz: float = DEFAULT_RATE_HZ) -> list[tuple[float, float, float]]:
        """(t, linear, angular) on a uniform grid. For --dry-run and footprint prediction."""
        n = max(1, int(round(self.duration_s * rate_hz)))
        out = []
        for i in range(n + 1):
            t = i / rate_hz
            lin, ang = self.at(t)
            out.append((t, lin, ang))
        return out


@dataclass(frozen=True)
class WaveformDecl:
    """One catalog entry, before it is built into a Program."""

    name: str
    kind: str
    channel: str
    role: str
    params: dict[str, Any]
    label: str = ""
    trim: bool = False
    motion: bool = True


# ---------------------------------------------------------------------------
# Segment assembly
# ---------------------------------------------------------------------------


class _Builder:
    """Accumulates held cells into a piecewise program."""

    def __init__(self) -> None:
        self.segments: list[ProtocolSegment] = []
        self.t = 0.0

    def hold(self, duration: float, linear: float, angular: float, label: str) -> None:
        if duration <= 0.0:
            return
        self.segments.append(
            ProtocolSegment(self.t, self.t + duration, linear, angular, label)
        )
        self.t += duration

    def coast(self, duration: float, label: str) -> None:
        self.hold(max(duration, MIN_COAST_S), 0.0, 0.0, label)


def _piecewise(segments: Sequence[ProtocolSegment]) -> Callable[[float], tuple[float, float]]:
    """Zero outside the program. Holding the last command past the end would tell the fit
    the robot was driven at full command while it sat on the floor."""
    bounds = [s.t0 for s in segments]

    def at(t: float) -> tuple[float, float]:
        if not segments or t < segments[0].t0 or t >= segments[-1].t1:
            return 0.0, 0.0
        lo, hi = 0, len(segments) - 1
        while lo < hi:
            mid = (lo + hi + 1) // 2
            if bounds[mid] <= t:
                lo = mid
            else:
                hi = mid - 1
        seg = segments[lo]
        return (seg.linear, seg.angular) if seg.t0 <= t < seg.t1 else (0.0, 0.0)

    return at


def _axes(channel: str, primary: float, secondary: float = 0.0) -> tuple[float, float]:
    """Map an amplitude onto (linear, angular) for the declared channel."""
    if channel == "linear":
        return primary, secondary
    if channel == "angular":
        return secondary, primary
    return primary, secondary


def cap_amplitudes(amplitudes: Sequence[float], cap: float | None) -> list[float]:
    """Scale a ladder so its peak lands on `cap`, rather than clamping to it.

    Clamping would collapse the top rungs onto one value and destroy the spacing the fit
    reads a gain slope from. Scaling keeps the ratios and just makes the whole ladder
    smaller, which is what the encoder-wheel rate limit actually requires.
    """
    values = [float(a) for a in amplitudes]
    if cap is None or not values:
        return values
    peak = max(abs(a) for a in values)
    if peak <= cap or peak <= 0.0:
        return values
    scale = cap / peak
    return [round(a * scale, 3) for a in values]


# ---------------------------------------------------------------------------
# Generators
# ---------------------------------------------------------------------------


def _build_step(d: WaveformDecl) -> Program:
    p = d.params
    amps = cap_amplitudes(p.get("amplitudes", [0.25, 0.5, 0.75, 1.0]), p.get("cap"))
    hold = max(float(p.get("hold_s", MIN_HOLD_S)), MIN_HOLD_S)
    coast = float(p.get("coast_s", MIN_COAST_S))
    shuttle = bool(p.get("shuttle", True))
    b = _Builder()
    for amp in amps:
        signs = (1.0, -1.0) if shuttle else (1.0,)
        for sign in signs:
            lin, ang = _axes(d.channel, sign * amp, sign * float(p.get("secondary", 0.0)))
            b.hold(hold, lin, ang, f"step_{amp:g}")
            b.coast(coast, "coast")
    return _finish(d, b)


def _build_staircase(d: WaveformDecl) -> Program:
    """Creep the command up until the wheels break static friction.

    Stage 2 reported a deadzone of 0.04 because 0.04 was the smallest command it tested and
    the robot already moved there. Starting at 0.01 is the point: several rungs must sit
    below the motion threshold or the crossing is an extrapolation.
    """
    p = d.params
    start = float(p.get("from", 0.01))
    stop = float(p.get("to", 0.10))
    step = float(p.get("step", 0.01))
    hold = max(float(p.get("hold_s", MIN_HOLD_S)), MIN_HOLD_S)
    b = _Builder()
    for sign in ((1.0, -1.0) if bool(p.get("both_directions", True)) else (1.0,)):
        level = start
        while level <= stop + 1e-9:
            lin, ang = _axes(d.channel, sign * level)
            b.hold(hold, lin, ang, f"staircase_{level:.3f}")
            level += step
        b.coast(MIN_COAST_S, "coast")
    return _finish(d, b)


def _build_coast(d: WaveformDecl) -> Program:
    """Accelerate, hold, then drop to zero instantly and let the robot roll to a stop.

    The drop must be instantaneous. A ramp down is a second input and the decel constant
    fitted through it is the ramp's, not the robot's.
    """
    p = d.params
    amp = float(p.get("amplitude", 0.6))
    hold = max(float(p.get("hold_s", MIN_HOLD_S)), MIN_HOLD_S)
    reps = int(p.get("reps", 10))
    shuttle = bool(p.get("shuttle", True))
    coast = float(p.get("coast_s", MIN_COAST_S))
    b = _Builder()
    for i in range(reps):
        sign = -1.0 if (shuttle and i % 2) else 1.0
        lin, ang = _axes(d.channel, sign * amp)
        b.hold(hold, lin, ang, "coast_entry")
        b.coast(coast, "coast_tail")
    return _finish(d, b)


def _build_grid(d: WaveformDecl) -> Program:
    """Cross linear against angular to separate the coupling terms.

    The angular list must span both signs. Steer-brake and angular droop flip with the turn
    direction while straight-line drift does not, so a grid that only turns one way cannot
    tell them apart and will fold the drift into the droop coefficient.
    """
    p = d.params
    lins = [float(v) for v in p.get("linear", [0.25, 0.5, 0.75, 1.0])]
    angs = cap_amplitudes(
        p.get("angular", [-0.3, -0.2, -0.1, 0.0, 0.1, 0.2, 0.3]), p.get("angular_cap")
    )
    hold = max(float(p.get("hold_s", MIN_HOLD_S)), MIN_HOLD_S)
    b = _Builder()
    for lin in lins:
        for ang in angs:
            b.hold(hold, lin, ang, f"grid_{lin:g}_{ang:g}")
            b.coast(MIN_COAST_S, "coast")
    return _finish(d, b)


def _build_prbs(d: WaveformDecl) -> Program:
    """Maximal-length 16-bit LFSR, taps 16/14/13/11.

    Reproduced bit-exactly from the web tool so runs recorded either side of the rewrite
    stay comparable. A 60 ms bit period covers roughly 0.2 to 8 Hz, which brackets both the
    58-78 ms plant time constants and the 33 ms frame period.
    """
    p = d.params
    amp = float(p.get("amplitude", 0.6))
    bit_ms = float(p.get("bit_ms", 60.0))
    duration = float(p.get("duration_s", 30.0))
    seed = int(p.get("seed", 0xACE1)) & 0xFFFF or 0xACE1
    ang_amp = float(p.get("angular_amplitude", amp))

    def bits(state: int, n: int) -> tuple[list[int], int]:
        out = []
        s = state
        for _ in range(n):
            bit = ((s >> 0) ^ (s >> 2) ^ (s >> 3) ^ (s >> 5)) & 1
            s = ((s >> 1) | (bit << 15)) & 0xFFFF
            out.append(1 if bit else -1)
        return out, s

    n = max(1, math.ceil(duration * 1000.0 / bit_ms))
    seq_a, state = bits(seed, n)
    # A second, decorrelated sequence for the angular channel of a combined run. Reusing one
    # sequence on both channels would make them perfectly correlated, and the coupling terms
    # would be unidentifiable no matter how long the run.
    seq_b, _ = bits((state * 1103515245 + 12345) & 0xFFFF or 0xBEEF, n)

    def at(t: float) -> tuple[float, float]:
        if t < 0.0 or t >= duration:
            return 0.0, 0.0
        i = min(n - 1, int(t * 1000.0 / bit_ms))
        if d.channel == "linear":
            return amp * seq_a[i], 0.0
        if d.channel == "angular":
            return 0.0, ang_amp * seq_a[i]
        return amp * seq_a[i], ang_amp * seq_b[i]

    return Program(
        name=d.name,
        kind=d.kind,
        channel=d.channel,
        role=d.role,
        duration_s=duration,
        at=at,
        segments=None,
        label=d.label,
        params=dict(p),
    )


def _build_chirp(d: WaveformDecl) -> Program:
    """Swept sine, logarithmic by default.

    A first-order model with tau 58 ms puts its corner near 2.7 Hz. A sweep through that
    corner is the frequency-domain cross-check on a time-constant fitted from steps.
    """
    p = d.params
    f0 = float(p.get("f0_hz", 0.2))
    f1 = float(p.get("f1_hz", 8.0))
    amp = float(p.get("amplitude", 0.4))
    duration = float(p.get("duration_s", 30.0))
    log_sweep = bool(p.get("log_sweep", True))

    def phase(t: float) -> float:
        if log_sweep and f0 > 0.0 and f1 > 0.0 and abs(f1 - f0) > 1e-9:
            k = math.log(f1 / f0) / duration
            return 2.0 * math.pi * f0 * (math.exp(k * t) - 1.0) / k
        return 2.0 * math.pi * (f0 * t + 0.5 * (f1 - f0) / duration * t * t)

    def at(t: float) -> tuple[float, float]:
        if t < 0.0 or t >= duration:
            return 0.0, 0.0
        return _axes(d.channel, amp * math.sin(phase(t)))

    return Program(d.name, d.kind, d.channel, d.role, duration, at, None, d.label, dict(p))


def _periodic(d: WaveformDecl, shape: Callable[[float], float]) -> Program:
    p = d.params
    freq = float(p.get("freq_hz", 1.0))
    amp = float(p.get("amplitude", 0.5))
    offset = float(p.get("offset", 0.0))
    if p.get("cycles") is not None and freq > 0.0:
        duration = float(p["cycles"]) / freq
    else:
        duration = float(p.get("duration_s", 20.0))

    def at(t: float) -> tuple[float, float]:
        if t < 0.0 or t >= duration:
            return 0.0, 0.0
        value = offset + amp * shape(freq * t)
        return _axes(d.channel, max(-1.0, min(1.0, value)))

    return Program(d.name, d.kind, d.channel, d.role, duration, at, None, d.label, dict(p))


def _build_sine(d: WaveformDecl) -> Program:
    return _periodic(d, lambda cycles: math.sin(2.0 * math.pi * cycles))


def _build_triangle(d: WaveformDecl) -> Program:
    # Constant slew between the peaks, so unlike a sine it excites the rate limit evenly
    # rather than concentrating the fastest command changes at the zero crossings.
    return _periodic(d, lambda cycles: 4.0 * abs(((cycles - 0.25) % 1.0) - 0.5) - 1.0)


def _build_manual(d: WaveformDecl) -> Program:
    """Emits zeros. The operator drives or pushes; the radio readback is the command log."""
    duration = float(d.params.get("duration_s", 60.0))
    return Program(
        name=d.name,
        kind=d.kind,
        channel=d.channel,
        role=d.role,
        duration_s=duration,
        at=lambda t: (0.0, 0.0),
        segments=None,
        label=d.label,
        params=dict(d.params),
    )


def _build_trim(d: WaveformDecl) -> Program:
    """Low-amplitude straight drive used to measure the arc, not to fit anything.

    Run before a session to find out whether a straight run will stay on the board, and to
    seed the drift parameter before any fit exists. Driven open loop here; the closed-loop
    correction lives in the CLI, which can see the jig's live stream.
    """
    p = d.params
    amp = float(p.get("amplitude", 0.15))
    duration = float(p.get("duration_s", 5.0))
    b = _Builder()
    b.hold(duration, amp, 0.0, "trim")
    b.coast(MIN_COAST_S, "coast")
    return _finish(d, b)


def _finish(d: WaveformDecl, b: _Builder) -> Program:
    return Program(
        name=d.name,
        kind=d.kind,
        channel=d.channel,
        role=d.role,
        duration_s=b.t,
        at=_piecewise(b.segments),
        segments=b.segments,
        label=d.label,
        params=dict(d.params),
    )


BUILDERS: dict[str, Callable[[WaveformDecl], Program]] = {
    "step": _build_step,
    "staircase": _build_staircase,
    "coast": _build_coast,
    "grid": _build_grid,
    "prbs": _build_prbs,
    "chirp": _build_chirp,
    "sine": _build_sine,
    "triangle": _build_triangle,
    "trim": _build_trim,
    "manual": _build_manual,
}


def build(decl: WaveformDecl) -> Program:
    try:
        builder = BUILDERS[decl.kind]
    except KeyError:
        raise ValueError(f"{decl.name}: no builder for kind {decl.kind!r}") from None
    return builder(decl)


# ---------------------------------------------------------------------------
# Catalog
# ---------------------------------------------------------------------------

_META_KEYS = {"kind", "channel", "role", "label", "trim", "motion"}


def load_catalog(path: Path | str) -> dict[str, WaveformDecl]:
    """Read a waveform TOML. `[defaults]` is merged under every entry."""
    path = Path(path)
    with open(path, "rb") as handle:
        data = tomllib.load(handle)

    defaults = dict(data.get("defaults", {}))
    out: dict[str, WaveformDecl] = {}
    for name, table in (data.get("waveform") or {}).items():
        if not isinstance(table, dict):
            raise ValueError(f"{path}: [waveform.{name}] is not a table")
        kind = str(table.get("kind", "")).strip()
        channel = str(table.get("channel", "")).strip()
        role = str(table.get("role", "fit")).strip()
        if kind not in WAVEFORM_KINDS:
            raise ValueError(f"{path}: [waveform.{name}] kind {kind!r} not in {WAVEFORM_KINDS}")
        if channel not in CHANNELS:
            raise ValueError(f"{path}: [waveform.{name}] channel {channel!r} not in {CHANNELS}")
        if role not in ROLES:
            raise ValueError(f"{path}: [waveform.{name}] role {role!r} not in {ROLES}")
        params = {k: v for k, v in defaults.items() if k not in _META_KEYS}
        params.update({k: v for k, v in table.items() if k not in _META_KEYS})
        out[name] = WaveformDecl(
            name=name,
            kind=kind,
            channel=channel,
            role=role,
            params=params,
            label=str(table.get("label", "")),
            trim=bool(table.get("trim", defaults.get("trim", False))),
            motion=bool(table.get("motion", True)),
        )
    if not out:
        raise ValueError(f"{path}: no [waveform.*] entries")
    return out
